src/storage-controller/src/lib.rs

```diff
@@ -807,7 +842,7 @@ where
         }
 
         // Register the tables all in one batch.
-        if !table_registers.is_empty() {
+        if !table_registers.is_empty() && !self.read_only {
```

Comment:
Maybe extend the comment here to explain that we can't register in read-only mode and that registering is deferred to when we transition out of it.